implementations/rust/ockam/ockam_vault/src/vault.rs

```
@@ -36,86 +35,62 @@ use ockam_node::compat::asynchronous::RwLock;
         }
     }
 
-    /// Serialize a vault to bytes which may later be restored using
-    /// `SoftwareVault::deserialize`.
-    #[cfg(feature = "storage")]
-    pub async fn serialize(&self) -> Vec<u8> {
-        serialize(&self.data).await
-    }
-
-    /// Load a vault from the serialized format produced by `SoftwareVault::serialize`.
-    #[cfg(feature = "storage")]
-    #[tracing::instrument(err, skip_all)]
-    pub fn deserialize(data: &[u8]) -> ockam_core::Result<Self> {
-        let data = deserialize(data).map_err(|e| {
-            tracing::error!("Data loaded from vault failed to parse: {:?}", e);
-            crate::VaultError::StorageError
-        })?;
-        Ok(Self { data })
-    }
-
     /// Same as ```Vault::new()```
     pub fn create() -> Self {
-        Self::new()
+        Self::new(None)
     }
-}
 
-#[derive(Debug, Eq, PartialEq, Clone)]
-#[cfg_attr(feature = "storage", derive(serde::Serialize, serde::Deserialize))]
-pub(crate) struct VaultEntry {
-    key_id: Option<String>,
-    key_attributes: SecretAttributes,
-    key: SecretKey,
-}
+    pub(crate) async fn preload_from_storage(&self, key_id: &KeyId) -> Result<()> {
+        if self.data.entries.read().await.contains_key(key_id) {
+            return Ok(());
+        }
 
-impl VaultEntry {
-    pub fn key_id(&self) -> &Option<String> {
-        &self.key_id
-    }
-    pub fn key_attributes(&self) -> SecretAttributes {
-        self.key_attributes
-    }
-    pub fn key(&self) -> &SecretKey {
-        &self.key
-    }
-}
+        let storage = match &self.storage {
+            Some(s) => s,
+            None => return Ok(()),
+        };
 
-impl VaultEntry {
-    pub fn new(key_id: Option<String>, key_attributes: SecretAttributes, key: SecretKey) -> Self {
-        VaultEntry {
-            key_id,
-            key_attributes,
-            key,
-        }
+        let entry = match storage.load(key_id).await {
+            Ok(e) => e,
+            Err(_) => return Ok(()),
```

Why is the error ignored?

---

The strategy is following:
We call Vault function (e.g. sign) with some key_id. We either:
1) Have corresponding key loaded into memory (Vault::data field)
2) Don't have key in the memory, but have it in storage
3) Don't have key at all

So, we always call preload_from_storage before working with the secret and the corresponding behaviour should be following:
1) Do nothing
2) Load key from the storage to the memory
3) Try to load key from the storage, fail silently

The calling function after calling preload_from_storage will try to get secret from the memory, and if it still not there - return an error, because it's 3)

---

Maybe worth mentioning that in the method's doc comment.